sql/lease.go

```diff
@@ -0,0 +1,546 @@
+// Copyright 2015 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License. See the AUTHORS file
+// for names of contributors.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package sql
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"math/rand"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/cockroachdb/cockroach/client"
+	"github.com/cockroachdb/cockroach/sql/parser"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/hlc"
+	"github.com/cockroachdb/cockroach/util/log"
+	"github.com/cockroachdb/cockroach/util/retry"
+)
+
+// TODO(pmattis): Periodically renew leases for tables that were used recently and
+// for which the lease will expire soon.
+
+const (
+	leaseDuration    = int64(5 * time.Minute)
+	minLeaseDuration = int64(time.Minute)
+)
+
+const (
+	leaseDescIDRowIndex = iota
+	leaseVersionRowIndex
+	leaseExpirationRowIndex
+	leaseNodeIDRowIndex
+)
+
+var (
+	leaseValTypes          []parser.Datum
+	leaseColIDToRowIndex   = make(map[ColumnID]int, 4)
+	errLeaseVersionChanged = errors.New("lease version changed")
+)
+
+func init() {
+	columns := [...]string{
+		leaseDescIDRowIndex:     "descID",
+		leaseVersionRowIndex:    "version",
+		leaseExpirationRowIndex: "expiration",
+		leaseNodeIDRowIndex:     "nodeID",
+	}
+	for i, name := range columns {
+		col, err := LeaseTable.FindColumnByName(name)
+		if err != nil {
+			log.Fatal(err)
+		}
+		if col.ID == 0 {
+			// This shouldn't happen, just being paranoid.
+			log.Fatalf("column %s invalid", col.Name)
+		}
+		leaseColIDToRowIndex[col.ID] = i
+
+		// Verify the column is at the expected location in the primary index.
+		index := -1
+		for i := range LeaseTable.PrimaryIndex.ColumnIDs {
+			if col.ID == LeaseTable.PrimaryIndex.ColumnIDs[i] {
+				index = i
+				break
+			}
+		}
+		if i != index {
+			log.Fatalf("unexpected index of \"%s\" in primary key: %d vs %d", name, i, index)
+		}
+	}
+
+	var err error
+	leaseValTypes, err = makeKeyVals(&LeaseTable, LeaseTable.PrimaryIndex.ColumnIDs)
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func nanosToTime(nanos int64) time.Time {
+	return time.Unix(nanos/1e9, nanos%1e9)
+}
+
+func makeLeaseKey(nodeID uint32, descID ID, version uint32, expiration int64) []byte {
```

Comment:
i think it was said elsewhere, but this seems like a reimplementation of some sql internals